src/actions/attack.cpp

```diff
@@ -1604,57 +1605,54 @@ int under_leadership(const unit &u, const map_location& loc, const_attack_ptr we
 	return leader_effect.get_composite_value();
 }
 
-int attack_type::specials_alignment(int combat_modifier, bool is_fearless) const
-{
-	if(!has_special_or_ability("alignment")){
-		return combat_modifier;
-	}
-	// if "alignment" attribute not eqauls to string in checking_alignment chain,
-	// or not "special_alignment special/abilities then alignment_type returned by default.
+static void alignment_list(unit_ability_list& abil_list, int& combat_modifier)
+{	//check if have abilities or specials alignment and remove alignment bad specified
 	const std::set<std::string> checking_alignment = {"neutral", "lawful", "chaotic", "liminal"};
-	unit_ability_list abil_list = get_specials_and_abilities("alignment");
 	for(unit_ability_list::iterator i = abil_list.begin(); i != abil_list.end();) {
 		if(checking_alignment.count((*i->ability_cfg)["alignment"].str()) == 0){
 			i = abil_list.erase(i);
 		} else {
 			++i;
 		}
 	}
+	//check if alignment of unit value can be or not used
+	for(const auto& i : abil_list) {
+		if(combat_modifier != 0 && (*i.ability_cfg)["no_alignment_type"].to_bool()) {
+			combat_modifier = 0;
+			break;
+		}
+	}
+}
+
+static int lawful_and_max_liminal_bonus(unit_ability_list& abil_list, const std::string& alignment, int bonus_value)
```

Comment:
Awesome work! But my bad, I think the name I originally proposed, `lawful_and_max_liminal_bonus`, is not a good name. It would be better to call it `get_combat_bonus`.